File: harness/python/run_script.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
HARNESS_DIR = ROOT / "harness"
sys.path.insert(0, str(ROOT / "ann" / "python"))

from algorithms import create_index
from ann_interface import NearestNeighbors
from script_loader import BuildOp, QueryOp, load_script

DEFAULT_ALGO = "brute_force"
DEFAULT_SCRIPT = HARNESS_DIR / "script.cases"


def run_script(path: Path, algo: str = DEFAULT_ALGO) -> None:
    index: NearestNeighbors = create_index(algo)

    for operation in load_script(path):
        if isinstance(operation, BuildOp):
            index.build(operation.points)
            print("build ok")
            continue

        results = index.query(operation.point, operation.k)
        print(f"query k {operation.k}")
        for neighbor in results:
            print(neighbor.index)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run ANN test script")
    parser.add_argument(
        "script",
        nargs="?",
        default=str(DEFAULT_SCRIPT),
        help="path to script.cases (default: harness/script.cases)",
    )
    parser.add_argument(
        "--algo",
        default=DEFAULT_ALGO,
        help="ANN algorithm to run (default: brute_force)",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    script_path = Path(args.script)
    if not script_path.is_absolute():
        script_path = ROOT / script_path

    run_script(script_path, args.algo)


if __name__ == "__main__":
    main()

File: harness/cpp/run_script.cpp
#include "algorithms.hpp"
#include "script_loader.hpp"

#include <iostream>
#include <string>

namespace {

constexpr const char* kDefaultAlgo = "brute_force";

void run_script(const std::string& path, const std::string& algo) {
    std::unique_ptr<NearestNeighbors> index = create_index(algo);

    for (const Operation& operation : load_script(path)) {
        if (std::holds_alternative<BuildOp>(operation)) {
            index->build(std::get<BuildOp>(operation).points);
            std::cout << "build ok\n";
            continue;
        }

        const QueryOp& query = std::get<QueryOp>(operation);
        const std::vector<Neighbor> results = index->query(query.point, query.k);
        std::cout << "query k " << query.k << '\n';
        for (const Neighbor& neighbor : results) {
            std::cout << neighbor.index << '\n';
        }
    }
}

struct RunnerOptions {
    std::string script_path = "harness/script.cases";
    std::string algo = kDefaultAlgo;
};

RunnerOptions parse_args(int argc, char* argv[]) {
    RunnerOptions options;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--algo") {
            if (i + 1 >= argc) {
                throw std::runtime_error("--algo requires a value");
            }
            options.algo = argv[++i];
            continue;
        }

        if (arg.rfind("--", 0) == 0) {
            throw std::runtime_error("unknown flag: " + arg);
        }

        options.script_path = arg;
    }

    return options;
}

}  // namespace

int main(int argc, char* argv[]) {
    try {
        const RunnerOptions options = parse_args(argc, argv);
        run_script(options.script_path, options.algo);
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << '\n';
        return 1;
    }

    return 0;
}
